# Week 3: Introduction to Data Science 📈
# Pre-module

Now that you are well-versed in a new language (Python), you are ready to take a deeper look into the data set collected by your colleagues. This week's modules will introduce you to data science tools that you will access and use through the Python language. 

In this pre-module activity you will learn how to:
1. Load data in a csv file to a pandas DataFrame using **read_csv()**
2. Understand the DataFrame data type and its shape/axes
3. Check for and remove rows with missing values using **isna()** and **drop()**/**dropna()**
2. Export a pandas DataFrame as a csv file using **to_csv()**


## Heart failure dataset

Recall from previous weeks that your research group is interested in identifying risk factors to death in heart failure patients. Your team has compiled clinical records of patients with a previous record of stage III/IV heart failure, which were provided by several physicians whose patients were willing to participate in the study, into **hf_data_raw.csv**. Take a few minutes to open this file in Excel and take note of the information collected.

Although your team of researchers have taken great care to provide physicians with clear instructions for recording patient data, sometimes mistakes find their way into the dataset or some parts of the dataset may not be useful for our purpose. With hundreds of patients participating in the study, it is impossible for you to sift through each patient's data manually to ensure the dataset is clean for analysis. Your first task is to take a look at the dataset as-is, identify any obvious issues, and make any modifications required. 

## Loading a csv file
First, we must be able to inspect the dataset using code, rather than in Excel or other software for managing tabular data. Although Excel is a useful tool for data organization and analysis, data scientists can encounter much larger datasets that call for faster, more powerful tools. One of the main tools we will be using in this course is the **pandas** library. 

The syntax for importing a Python library, which you may have seen in weeks 1 and 2, is:
```
import <library name> as <alias>
```
The alias is not necessary, but using shorthands for long library names help improve readability, as we will call upon these libraries very frequently throughout our exploration. Some aliases are also widely accepted in industry and the computing community to facilitate better collaboration and communication. For example, *pd* is the widely accepted alias for pandas.

**Q1.** Let's import the pandas library with ```pd``` as the alias.
<span style="background-color: #FFD700">**Complete the code below to import the pandas library with *pd* as the alias**</span>

In [1]:
# TODO: Write your code here

The read_csv() function in pandas takes in a csv file and returns a DataFrame. A DataFrame is a data type for tabular data. 

read_csv() has one mandatory parameter: the path to the csv file you wish to convert to a DataFrame. Ensure the file hf_data.csv is uploaded to JupyterHub in the same folder that this module is in, if you have not already. Assuming that this file is in the same folder as this Jupyter Notebook, the filepath should be the name of the dataset file as a string (i.e. wrap the name in quotation marks).

| Function | Input parameters | Output | Syntax |
| --- | --- | --- | --- |
| read_csv() | filepath | pandas DataFrame | read_csv(filepath) |

**Q2.** Try it out- read hf_data_raw.csv in as a pandas DataFrame.

<span style="background-color: #FFD700">**Complete the code below.**</span>

In [ ]:
df = # TODO: Complete this line

# Let's see what the DataFrame looks like
df

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
298  62.0        0                        61         1                 38   
299  55.0        0                      1820         0                 38   
300  45.0        0                      2060         1                 60   
301  45.0        0                      2413         0                 38   
302  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1.0  265000.00               1.9           130    1   
1                    0.0  263358.03               1.1           136    1   
2                    0.0  162000.00               1.3           129    1   
3                    0.0  210000.00               1.9           137    1   
4                    0.0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
298                  1.0  155000.00               1.1           143    1   
299                  0.0  270000.00               1.2           139    0   
300                  0.0  742000.00               0.8           138    0   
301                  0.0  140000.00               1.4           140    1   
302                  0.0  395000.00               1.6           136    1   

     smoking  time  DEATH_EVENT  
0        0.0     4            1  
1        0.0     6            1  
2        1.0     7            1  
3        0.0     7            1  
4        0.0     8            1  
..       ...   ...          ...  
298      1.0   270            0  
299      0.0   271            0  
300      0.0   278            0  
301      1.0   280            0  
302      1.0   285            0  

[303 rows x 13 columns]

Great! You should be able to see some rows and columns of the DataFrame. You'll notice that this dataset is probably too big to show here, hence the ellipses. Other observations about a DataFrame:

1. DataFrame column labels correspond to the first row of the csv file, and are strings (text) in this case
2. DataFrame rows are numbered 0...num rows - 1 (i.e. they are 0-indexed)

There is a way to find out the number of rows and columns without printing df and observing the last row index. These are important parameters because if we want to access particular rows and/or columns later on, we need to have a sense of what is the maximum index we can access without going out of bounds (which will throw an error).

<span style="background-color: #FFD700">**Run the code below**</span>

In [16]:
print("Shape: ", df.shape)
rows = df.shape[0]
cols = df.shape[1]
print("Number of rows: ", rows)
print("Number of columns: ", cols)

Shape:  (303, 13)
Number of rows:  303
Number of columns:  13


<span style="background-color: #AFEEEE">**axis:**</span> a dimension of the DataFrame. Since a DataFrame is 2-dimensional (has rows and columns, like a matrix), there are 2 axes, 0 and 1.

**Q3.** For this dataset, which axis (0: rows, 1: cols) is features? Which axis is samples (individual patients)?

<span style="background-color: #FFD700">**Provide your answer here**</span>

We can use `head(n)` and `tail(n)` to display the first and last n rows of the dataset, respectively. 

<span style="background-color: #FFD700">**Run the two code cells below.**</span>

In [ ]:
df.head(5)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                  1.0  265000.00               1.9           130    1   
1                  0.0  263358.03               1.1           136    1   
2                  0.0  162000.00               1.3           129    1   
3                  0.0  210000.00               1.9           137    1   
4                  0.0  327000.00               2.7           116    0   

   smoking  time  DEATH_EVENT  
0      0.0     4            1  
1      0.0     6            1  
2      1.0     7            1  
3      0.0     7            1  
4      0.0     8            1

In [ ]:
df.tail(3)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
298  45.0        0                      2060         1                 60   
299  45.0        0                      2413         0                 38   
300  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
298                  0.0   742000.0               0.8           138    0   
299                  0.0   140000.0               1.4           140    1   
300                  0.0   395000.0               1.6           136    1   

     smoking  time  DEATH_EVENT  
298      0.0   278            0  
299      1.0   280            0  
300      1.0   285            0

As we can see, a pandas **DataFrame** has rows and columns of data; it presents data as a table. Each row and column is a pandas **Series**, which is basically like a list (covered in week 1). However, these are different from basic Python lists in that pandas provides us functions that can be specifically called on Series. Likewise, there are functions that we can call specifically on DataFrames. 

Remember from week 1 that we can access elements of an array by indexing. Now that we have a DataFrame, which is essentialy a table of elements, how do we access rows, columns, and individual elements?

To get a whole column, we do ```df['column name']```. This is the same syntax you saw from week 2, when you learned about dictionaries. 
To get a whole row, we do ```df.iloc[row index]```. ```iloc``` is a function that works on a pandas DataFrame that allows us to access parts of the DataFrame by index. Aside from having to use ```iloc```, it works the same as indexing into an array. 

<span style="background-color: #FFD700">**Run the code cell below.**</span> **Look the output of the 5th row printed below, and the last row in the output of df.head() above, to see that they are referring to the same row.**

In [ ]:
# print the age column
print("Print age column:")
print(df['age'])
# print the 5th row (index 4)
print("\nPrint the 4th row:")
print(df.iloc[4])

Print age column:
0      75.0
1      55.0
2      65.0
3      50.0
4      65.0
       ... 
296    62.0
297    55.0
298    45.0
299    45.0
300    50.0
Name: age, Length: 299, dtype: float64

Print the 4th row:
age                             65.0
anaemia                          1.0
creatinine_phosphokinase       160.0
diabetes                         1.0
ejection_fraction               20.0
high_blood_pressure              0.0
platelets                   327000.0
serum_creatinine                 2.7
serum_sodium                   116.0
sex                              0.0
smoking                          0.0
DEATH_EVENT                      1.0
Name: 4, dtype: float64


## Data cleaning

Let's think about what data cleaning involves. What needs to be "cleaned" will vary from dataset to dataset, but here are a few good starting points.

The most obvious thing we should look for are rows with missing values. Either the data collector (physicians in this case) has made an error or the information was not available. In a DataFrame, an entry at the i'th row and j'th column is empty or missing if it is **NaN** (Not a Number). Missing values can cause issues later when we compute statistics or apply machine learning models. For this exercise we will remove rows with missing values, but it is possible to retain such rows through data imputation (see Further Reading at the end of this module).

The DataFrame function **isna** returns a DataFrame with boolean values indicating whether the entry at that position is missing or not. It converts every data point into a boolean, so the resulting shape should be the same as your DataFrame.

| Function | Input parameters | Output | Syntax |
| --- | --- | --- | --- |
| isna() | n/a | A DataFrame of boolean values (whether the element is missing or not) | df.isna() |

<span style="background-color: #FFD700">**Run the code cell below**</span>

In [5]:
df.isna()

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    False    False                     False     False              False   
1    False    False                     False     False              False   
2    False    False                     False     False              False   
3    False    False                     False     False              False   
4    False    False                     False     False              False   
..     ...      ...                       ...       ...                ...   
298  False    False                     False     False              False   
299  False    False                     False     False              False   
300  False    False                     False     False              False   
301  False    False                     False     False              False   
302  False    False                     False     False              False   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium    sex  \
0                  False      False             False         False  False   
1                  False      False             False         False  False   
2                  False      False             False         False  False   
3                  False      False             False         False  False   
4                  False      False             False         False  False   
..                   ...        ...               ...           ...    ...   
298                False      False             False         False  False   
299                False      False             False         False  False   
300                False      False             False         False  False   
301                False      False             False         False  False   
302                False      False             False         False  False   

     smoking   time  DEATH_EVENT  
0      False  False        False  
1      False  False        False  
2      False  False        False  
3      False  False        False  
4      False  False        False  
..       ...    ...          ...  
298    False  False        False  
299    False  False        False  
300    False  False        False  
301    False  False        False  
302    False  False        False  

[303 rows x 13 columns]

We can't possibly look through every row in this format. For all we know, maybe there aren't any missing values in the entire dataset. We want to see only those rows with missing values.

The function **any()** returns whether any element in the set is True. This function can be called on an axis (row or column) in a DataFrame. If we call any() on the column axis, the function will check every column of a row and return True for that row if any of the columns were True. This would be done for all rows, so that the final result is a True/False value for each row.

| Function | Input parameters | Output | Syntax |
| --- | --- | --- | --- |
| any() | axis | The DataFrame, with the given rows/columns dropped. | any(axis) |

By using ```any()```, we reduce the above result to a single dimension. Reducing the dimensionality in this case helps us extract just the information we need.


<span style="background-color: #FFD700">**Run the code below to see which rows have missing values.**</span>

In [6]:
list_nan = df.isna().any(axis=1)
list_nan

0      False
1      False
2      False
3      False
4      False
       ...  
298    False
299    False
300    False
301    False
302    False
Length: 303, dtype: bool

**Q4.** We would like to see which columns have missing values, using a similar method as above with ```isna()``` and ```any()```.

<span style="background-color: #FFD700">**Complete the code below. You will only have to change the argument in ```any()```.**</span>

In [7]:
df. # TODO: complete this code

age                         False
anaemia                     False
creatinine_phosphokinase    False
diabetes                    False
ejection_fraction           False
high_blood_pressure          True
platelets                   False
serum_creatinine             True
serum_sodium                False
sex                         False
smoking                      True
time                        False
DEATH_EVENT                 False
dtype: bool

**Q5.** We can't possibly look through all 303 rows in ```list_nan``` above. But we can leverage a pandas feature called **boolean indexing**, which lets us **mask** (filter) a DataFrame with a list of booleans to return just the samples where the mask is True. We do this by using the boolean list as we would use an integer to index into a DataFrame, as such: ```data_frame[boolean_list]```.

<span style="background-color: #FFD700">**Complete the code below using list_nan to return just the rows with NaN values.** </span>

In [8]:
# TODO: Write your code here

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
36   55.0        0                        94         1                 50   
131  52.0        1                       172         0                 38   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
36                   NaN   188000.0               1.0           140    1   
131                  0.0   293000.0               NaN           137    1   

     smoking  time  DEATH_EVENT  
36       NaN    29            1  
131      0.0    11            1

Great! Now we can easily see that there are two patients with missing values; patient 36 is missing high blood pressure and smoking information, and patient 131 is missing their serum_creatinine measurement. How do we remove these patients from the dataset? Like most concepts we have covered so far, there are multiple ways to achieve the desired output.

There are two main ways to drop rows in pandas: ```drop()``` and ```dropna()```. We will discuss drop(), but you can read about dropna() in Further Reading.

### drop()
This function lets you drop rows by index. We figured out above that patients with index 36 and 131 need to be dropped.  

| Function | Input parameters | Output | Syntax |
| --- | --- | --- | --- |
| drop() | labels (index or column labels to drop), axis | The DataFrame, with the given rows/columns dropped. | df.drop(labels, axis) |

* labels: The row indices/column names indicating what you want to drop. If you want to drop multiple rows, the value you pass in for labels should be a list. In this case your list should contain the indices 36 and 131. 
* axis: The axis to drop from. If 0, rows indicated by labels will be dropped. If 1, columns indicated by labels will be dropped. By default, axis is 0 (so if you do not specify this parameter when you call drop(), the function will drop rows.)

**Q6.** Let's remove rows 36 and 131 and assign it to the variable ```df_drop```. You will need to call ```drop()``` on the variable ```df``` with the correct arguments.

<span style="background-color: #FFD700">**Complete the code below to remove rows 36 and 131 and assign it to df_drop.**</span>

In [9]:
df_drop = # TODO: Complete this line
df_drop

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
298  62.0        0                        61         1                 38   
299  55.0        0                      1820         0                 38   
300  45.0        0                      2060         1                 60   
301  45.0        0                      2413         0                 38   
302  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1.0  265000.00               1.9           130    1   
1                    0.0  263358.03               1.1           136    1   
2                    0.0  162000.00               1.3           129    1   
3                    0.0  210000.00               1.9           137    1   
4                    0.0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
298                  1.0  155000.00               1.1           143    1   
299                  0.0  270000.00               1.2           139    0   
300                  0.0  742000.00               0.8           138    0   
301                  0.0  140000.00               1.4           140    1   
302                  0.0  395000.00               1.6           136    1   

     smoking  time  DEATH_EVENT  
0        0.0     4            1  
1        0.0     6            1  
2        1.0     7            1  
3        0.0     7            1  
4        0.0     8            1  
..       ...   ...          ...  
298      1.0   270            0  
299      0.0   271            0  
300      0.0   278            0  
301      1.0   280            0  
302      1.0   285            0  

[301 rows x 13 columns]

Great! You have successfully removed rows with empty values.

## Exporting a DataFrame as csv
Now that we have our clean (as far as we can see) dataset, we will convert it back to a .csv file. We do this to save it in an easily shareable format which we can load into any notebook session in the future. You will be using this clean dataset for your main module during tutorial. 
This is achieved using the DataFrame **to_csv()** function. The index parameter specifies whether to save the row indices (0..max) as the first column. The default is True, but we will set this to False as this extra column is redundant for our purpose.

| Function | Input parameters | Output | Syntax |
| --- | --- | --- | --- |
| to_csv() | filename, index | n/a; saves the file to the same that your notebook is in | df.to_csv() |

* filename: The name that you want the output file to be saved as. This should be a string. For example, ```"output.csv"```.
* index: Specifies whether to save the row indices (0..max) as the first column. The default is True.

In [15]:
df_drop.to_csv('hf_data_pre_student.csv', index=False)

Check to see that this new csv file appears in your JupyterHub, in the same folder that this notebook is in.

## Conclusion
In this pre-module activity you will learn how to:
1. Load data in a csv file to a pandas DataFrame using **read_csv()**
2. Understand the DataFrame data type and its shape/axes
3. Check for and remove rows with missing values using **isna()** and **drop()**/**dropna()**
2. Export a pandas DataFrame as a csv file using **to_csv()**

## Further reading
* Data imputation: https://medium.com/godlessindian/imputation-methods-in-data-preprocessing-b225f7456de1 
* Pandas **isna** documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isna.html
    * Note: **isnull** is an alias of **isna** and can be used interchangeably.
* **drop_na()**
    * As an alternative of drop(), dropna() condenses the steps to drop rows with empty values. We can specify the scheme through the **how** parameter. Options for this parameter include:
        * ‘any’ : If at least one value is empty in the row/col, drop that row/col.
        * ‘all’ : If all values are empty the row/col, drop that row/col.
    * We should also specify the axis that we want this dropping scheme to apply to. We want to drop *rows* with empty values, so we would specify ```axis = 0``` in the parameters. 
    * For our example, ```df_dropna = df.dropna(how='any')``` or ```df.dropna(axis=0, how='any')``` would work instead of the code we wrote for ```drop()```.
    * Documentation for this function: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html#pandas.DataFrame.dropna